src/build.rs

```diff
@@ -0,0 +1,121 @@
+use std;
+use super::{
+	CREATE_SYMBOL,
+	CALL_SYMBOL,
+	optimize,
+	pack_instance,
+	ununderscore_funcs,
+	externalize_mem,
+	shrink_unknown_stack,
+	inject_runtime_type,
+	PackingError,
+	OptimizerError
+	};
+use parity_wasm;
+use parity_wasm::elements;
+
+
+#[derive(Debug)]
+pub enum Error {
+	Encoding(elements::Error),
+	Packing(PackingError),
+	NoCreateSymbolFound,
+	Optimizer,
+}
+
+impl From<OptimizerError> for Error {
+	fn from(_err: OptimizerError) -> Self {
+		Error::Optimizer
+	}
+}
+
+impl From<PackingError> for Error {
+	fn from(err: PackingError) -> Self {
+		Error::Packing(err)
+	}
+}
+
+#[derive(Debug, Clone, Copy)]
+pub enum Target {
+	Emscripten,
+	Unknown,
+}
+
+impl std::fmt::Display for Error {
+	fn fmt(&self, f: &mut std::fmt::Formatter) -> Result<(), std::fmt::Error> {
+		use self::Error::*;
+		match *self {
+			Encoding(ref err) => write!(f, "Encoding error ({})", err),
+			Optimizer => write!(f, "Optimization error due to missing export section. Pointed wrong file?"),
+			Packing(ref e) => write!(f, "Packing failed due to module structure error: {}. Sure used correct libraries for building contracts?", e),
+			NoCreateSymbolFound => write!(f, "Packing failed: no \"{}\" symbol found?", CREATE_SYMBOL),
+		}
+	}
+}
+
+fn has_ctor(module: &elements::Module) -> bool {
+	if let Some(ref section) = module.export_section() {
+		section.entries().iter().any(|e| CREATE_SYMBOL == e.field())
+	} else {
+		false
+	}
+}
+
+pub fn build(
+	mut module: elements::Module,
+	constructor: bool,
+	target: Target,
+	runtime_type: Option<&[u8]>,
+	runtime_version: Option<u32>,
+	mut public_api_entries: Vec<&str>,
```

Comment:
this is bad part of the signature for public api

`&[&str]` would be ok

Reply:
can you explain? I'd better to use [u8; 4] or u32, not &str

Reply:
I'd like to use the type system soundness here. So I'd better use [u8; 4] here

Reply:
what is `[u8; 4]`? how it is related to `Vec<&str>`?

Reply:
so for user it will be more descriptive API which could help to avoid the runtime error

Reply:
how can you put a string in `[u8; 4]` ?

Reply:
Cause for runtime_type we use exactly 4 bytes. Would be nice to declare it as it is 🤗

Reply:
the conversation is about string `public_api_entries` argument

Reply:
ah, ok 🙈